Lib/dataclasses.py

```diff
@@ -411,36 +416,77 @@ def _fields_in_init_order(fields):
             )
 
-
 def _tuple_str(obj_name, fields):
     # Return a string representing each field of obj_name as a tuple
     # member.  So, if fields is ['x', 'y'] and obj_name is "self",
-    # return "(self.x,self.y)".
+    # return "(self._field_x,self._field_y)".
 
     # Special case for the 0-tuple.
     if not fields:
         return '()'
     # Note the trailing comma, needed if this turns out to be a 1-tuple.
-    return f'({",".join([f"{obj_name}.{f.name}" for f in fields])},)'
+    return f'({",".join([f"{obj_name}.{f._placeholder}" for f in fields])},)'
 
 
-def _create_fn(name, args, body, *, globals=None, locals=None,
-               return_type=MISSING):
+_code_cache = {}
+_code_cache_hits = 0
+
+def _create_fn(name, args, body, *, globals=None, locals=None):
     # Note that we may mutate locals. Callers beware!
     # The only callers are internal to this module, so no
     # worries about external callers.
     if locals is None:
         locals = {}
-    return_annotation = ''
-    if return_type is not MISSING:
-        locals['_return_type'] = return_type
-        return_annotation = '->_return_type'
     args = ','.join(args)
     body = '\n'.join(f'  {b}' for b in body)
 
     # Compute the text of the entire function.
-    txt = f' def {name}({args}){return_annotation}:\n{body}'
+    txt = f' def {name}({args}):\n{body}'
+
+    # exec is *really* slow, so we want to avoid using it whenever possible.
+    # Turns out, if we already have a code object that is the exact same as the
+    # one we want to create (except for the actual names of the fields), it's
+    # twice as fast to just copy the other code object, fix up the names, and
+    # create a function using the types.FunctionType constructor.
+
+    # In order for this to work, we need to be able to detect the names of the
+    # fields from the source. For this reason, it's *crucial* that anything
+    # building source strings uses field._placeholder, *not* field.name!
+
+    # This trick requires two passes:
+
+    # - Strip out the field names from the source text, and replace them with
+    #   something else (like numbers). This is used to search for (or create)
+    #   cached code with the same structure. This happens in _extract_fields.
 
+    # - Make a copy of the cached code object, and fix up all of the stripped
+    #   names in the consts, names, and varnames. This happens in _patch_fields.
+
+    # This is surprisingly effective, since we generate lots of functions that
+    # differ *only* in the names of their fields (we avoid something like 95% of
+    # the exec calls in test_dataclasses.py). Credit for this neat idea goes to
+    # https://github.com/dabeaz/dataklasses.
+
+    patches, key = _extract_fields(txt)
+    if key in _code_cache:
+        global _code_cache_hits
+        _code_cache_hits += 1
+        code = _code_cache[key]
+    else:
+        code = _code_cache[key] = _create_code(name, key, globals, locals)
```

Comment:
I think that this will make `regtest` fail. We have a similar pattern in `typing.py`: https://github.com/python/cpython/blame/8dd2766d99f8f51ad62dc0fde8282483590c6cd0/Lib/typing.py#L2550-L2603

And we need to call `clear_overloads` to clear module level cache. Otherwise we get "Env has changed" error.

Reply:
Hm, I'm not sure I understand. I don't see any test failures locally, even when testing for refleaks.

I also only see one use of `clear_overloads` in the entire code base (as part of a normal test to make sure that calling it works).

Reply:
Ok, if `refleak` test passes, then it is fine :) Thanks for double checking!